AppKit/CPWindow/CPWindow.j

```diff
@@ -3844,6 +3819,393 @@ var interpolate = function(fromValue, toValue, progress)
 
 @end
 
+@implementation CPWindow (TrackingAreaAdditions)
+
+- (void)_addTrackingAreaView:(CPView)aView
+{
+    var trackingAreas = [aView trackingAreas];
+    
+    for (var i = 0; i < trackingAreas.length; i++)
+        [self _addTrackingArea:trackingAreas[i]];
+}
+
+- (void)_removeTrackingAreaView:(CPView)aView
+{
+    var trackingAreas = [aView trackingAreas];
+    
+    for (var i = 0; i < trackingAreas.length; i++)
+        [self _removeTrackingArea:trackingAreas[i]];
+}
+
+- (void)_addTrackingArea:(CPTrackingArea)trackingArea
+{
+    var trackingAreaView = [trackingArea view];
+    
+    if (![_trackingAreaViews containsObject:trackingAreaView])
+        [_trackingAreaViews addObject:trackingAreaView];
+    
+    // If CPTrackingAssumeInside option is set, put the tracking area in the _mouseEnteredStack
+    
+    if ([trackingArea options] & CPTrackingAssumeInside)
+        [_mouseEnteredStack addObject:trackingArea];
+}
+
+- (void)_removeTrackingArea:(CPTrackingArea)trackingArea
+{
+    // If mouse is in the tracking area, we remove it from the stack to avoid to fire a future mouseExited event
+    
+    [_mouseEnteredStack removeObjectIdenticalTo:trackingArea];
+    
+    var trackingAreaView = [trackingArea view];
+    
+    [_trackingAreaViews removeObjectIdenticalTo:trackingAreaView];
+}
+
+- (void)_handleTrackingAreaEvent:(CPEvent)anEvent
```

Comment:
> shouldn't we apply the following general policy : each layer of code should use lower layers respecting APIs (this is for readability), and should optimize its own implementation

In general that's what already happens in Cappuccino. But when an Objective-J class is toll-free bridged with a native object (like strings and arrays), I usually use the native syntax mostly because it's more concise. For example, I would much rather use `anArray[i]` than `[anArray objectAtIndex:i]`. Apple agrees with me! They added this more concise array indexing syntax to Objective-C and it is now the preferred way to access array elements. So why should we go backwards to a less concise syntax?

Similarly, to me it's pointless to use `for (var i = 0, count = [array count]; i < count; i++)` instead of the more concise `for (var i = 0; i < array.length; i++)`. The extra `count` variable just obscures what's going on and makes it harder to visually parse the statement. And in Objective-C, where they have properties, `for (int i = 0; i < array.count; i++)` is exactly the syntax you would use (if you weren't going to use fast enumeration). Again, why would we use a less concise syntax? The same thing goes for strings.

We should remember that Objective-J is still Javascript, and there is no shame in using Javascript instead of Objective-J when it's possible.

As for the performance factor, your tests confirm exactly what I said: `containsObject` is a killer. As both cacaodev and I pointed out, you should have been using `containsObjectIdenticalTo` anyway, because you want to compare object identities, not logical equivalence. Any time you would use === to compare individual objects, you should use `containsObjectIdenticalTo`. So that's something you should fix in all of this code.

As for factoring this method, because of the optimizations that Martin has made to method dispatch, the overhead of three more methods calls is effectively zero.